ERROR : INVALID_ARGUMENT jobInternalError

From time to time, we have a jobInternalError 

The jobInternalError is a broad error message, and it can be triggered by various underlying issues.
It can be a temporary network error on our side, or a temporary issue on GCP but it's a problem to have to manually restart the connector each time this happen.

After the connector restart everything is fine and run smoothly again 


If this error is caused by temporary general issues, couldn't be better to consider it as a Retryable error instead of just crashing the connector task ?


The error from GCP : 
```
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "The job encountered an internal error during execution and was unable to complete successfully.",
    "reason" : "jobInternalError"
  } ],
  "message" : "The job encountered an internal error during execution and was unable to complete successfully.",
  "status" : "INVALID_ARGUMENT"
}
```


The stacktrace : 
```
ERROR WorkerSinkTask{id=connector_sink_bigquery-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted. Error: A write thread has failed with an unrecoverable error (org.apache.kafka.connect.runtime.WorkerSinkTask:566)
com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException: A write thread has failed with an unrecoverable error
Caused by: The job encountered an internal error during execution and was unable to complete successfully.
	at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.lambda$maybeThrowEncounteredError$0(KCBQThreadPoolExecutor.java:101)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.maybeThrowEncounteredError(KCBQThreadPoolExecutor.java:100)
	at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.writeSinkRecords(BigQuerySinkTask.java:251)
	at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.put(BigQuerySinkTask.java:304)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:546)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:326)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:228)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:196)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:184)
	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.google.cloud.bigquery.BigQueryException: The job encountered an internal error during execution and was unable to complete successfully.
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.translate(HttpBigQueryRpc.java:115)
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.getQueryResults(HttpBigQueryRpc.java:643)
	at com.google.cloud.bigquery.BigQueryImpl$36.call(BigQueryImpl.java:1393)
	at com.google.cloud.bigquery.BigQueryImpl$36.call(BigQueryImpl.java:1388)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.bigquery.BigQueryRetryHelper.run(BigQueryRetryHelper.java:86)
	at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:49)
	at com.google.cloud.bigquery.BigQueryImpl.getQueryResults(BigQueryImpl.java:1387)
	at com.google.cloud.bigquery.BigQueryImpl.getQueryResults(BigQueryImpl.java:1371)
	at com.google.cloud.bigquery.Job$1.call(Job.java:338)
	at com.google.cloud.bigquery.Job$1.call(Job.java:335)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.bigquery.BigQueryRetryHelper.run(BigQueryRetryHelper.java:86)
	at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:49)
	at com.google.cloud.bigquery.Job.waitForQueryResults(Job.java:334)
	at com.google.cloud.bigquery.Job.getQueryResults(Job.java:295)
	at com.google.cloud.bigquery.BigQueryImpl.queryRpc(BigQueryImpl.java:1335)
	at com.google.cloud.bigquery.BigQueryImpl.query(BigQueryImpl.java:1283)
	at com.wepay.kafka.connect.bigquery.MergeQueries.mergeFlush(MergeQueries.java:153)
	at com.wepay.kafka.connect.bigquery.MergeQueries.lambda$mergeFlush$1(MergeQueries.java:130)
	... 3 more
```

--- a/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java
+++ b/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java
@@ -27,6 +27,7 @@ import com.google.cloud.bigquery.QueryJobConfiguration;
 import com.google.cloud.bigquery.Schema;
 import com.google.cloud.bigquery.TableId;
 import com.google.common.annotations.VisibleForTesting;
+import com.wepay.kafka.connect.bigquery.config.BigQuerySinkConfig;
 import com.wepay.kafka.connect.bigquery.config.BigQuerySinkTaskConfig;
 import com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException;
 import com.wepay.kafka.connect.bigquery.exception.ExpectedInterruptException;
@@ -35,7 +36,6 @@ import com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor;
 import com.wepay.kafka.connect.bigquery.write.batch.MergeBatches;
 import com.wepay.kafka.connect.bigquery.write.row.BigQueryErrorResponses;
 import org.apache.kafka.connect.errors.ConnectException;
-import org.apache.kafka.connect.errors.RetriableException;
 import org.apache.kafka.connect.sink.SinkTaskContext;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -49,6 +49,8 @@ import static com.wepay.kafka.connect.bigquery.utils.TableNameUtils.destTable;
 import static com.wepay.kafka.connect.bigquery.utils.TableNameUtils.intTable;
 
 public class MergeQueries {
+
+  private static final int WAIT_MAX_JITTER = 1000;
   public static final String INTERMEDIATE_TABLE_KEY_FIELD_NAME = "key";
   public static final String INTERMEDIATE_TABLE_VALUE_FIELD_NAME = "value";
   public static final String INTERMEDIATE_TABLE_ITERATION_FIELD_NAME = "i";
@@ -62,6 +64,8 @@ public class MergeQueries {
   private final boolean insertPartitionTime;
   private final boolean upsertEnabled;
   private final boolean deleteEnabled;
+  private final int bigQueryRetry;
+  private final long bigQueryRetryWait;
   private final MergeBatches mergeBatches;
   private final KCBQThreadPoolExecutor executor;
   private final BigQuery bigQuery;
@@ -78,9 +82,11 @@ public class MergeQueries {
       config.getKafkaKeyFieldName().orElseThrow(() ->
           new ConnectException("Kafka key field must be configured when upsert/delete is enabled")
       ),
-      config.getBoolean(config.BIGQUERY_PARTITION_DECORATOR_CONFIG),
-      config.getBoolean(config.UPSERT_ENABLED_CONFIG),
-      config.getBoolean(config.DELETE_ENABLED_CONFIG),
+      config.getBoolean(BigQuerySinkConfig.BIGQUERY_PARTITION_DECORATOR_CONFIG),
+      config.getBoolean(BigQuerySinkConfig.UPSERT_ENABLED_CONFIG),
+      config.getBoolean(BigQuerySinkConfig.DELETE_ENABLED_CONFIG),
+      config.getInt(BigQuerySinkConfig.BIGQUERY_RETRY_CONFIG),
+      config.getLong(BigQuerySinkConfig.BIGQUERY_RETRY_WAIT_CONFIG),
       mergeBatches,
       executor,
       bigQuery,
@@ -94,6 +100,8 @@ public class MergeQueries {
                boolean insertPartitionTime,
                boolean upsertEnabled,
                boolean deleteEnabled,
+               int bigQueryRetry,
+               long bigQueryRetryWait,
                MergeBatches mergeBatches,
                KCBQThreadPoolExecutor executor,
                BigQuery bigQuery,
@@ -103,6 +111,8 @@ public class MergeQueries {
     this.insertPartitionTime = insertPartitionTime;
     this.upsertEnabled = upsertEnabled;
     this.deleteEnabled = deleteEnabled;
+    this.bigQueryRetry = bigQueryRetry;
+    this.bigQueryRetryWait = bigQueryRetryWait;
     this.mergeBatches = mergeBatches;
     this.executor = executor;
     this.bigQuery = bigQuery;
@@ -150,16 +160,19 @@ public class MergeQueries {
       boolean success = false;
       while (!success) {
         try {
+          if (attempt > 0) {
+            SleepUtils.waitRandomTime(this.bigQueryRetryWait, WAIT_MAX_JITTER);
+          }
           bigQuery.query(QueryJobConfiguration.of(mergeFlushQuery));
           success = true;
         } catch (BigQueryException e) {
-          if (BigQueryErrorResponses.isCouldNotSerializeAccessError(e)) {
-            attempt++;
-            if (attempt == 30) {
-              throw new BigQueryConnectException("Failed to merge rows to destination table `" + destinationTable + "` within " + attempt 
-                                                  + "  attempts due to BQ write serialization error.", e);
-            }
-            SleepUtils.waitRandomTime(10000, 20000);
+          if (attempt >= bigQueryRetry) {
+            throw new BigQueryConnectException("Failed to merge rows to destination table `" + destinationTable + "` within " + attempt
+              + "  attempts.", e);
+          } else if (BigQueryErrorResponses.isCouldNotSerializeAccessError(e)) {
+            logger.warn("Serialize access error while merging from {} to {}, retry attempt {}", intermediateTable, destinationTable, ++attempt);
+          } else if (BigQueryErrorResponses.isJobInternalError(e)) {
+            logger.warn("Job internal error while merging from {} to {}, retry attempt {}", intermediateTable, destinationTable, ++attempt);
           } else {
             throw e;
           }

--- a/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/SleepUtils.java
+++ b/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/SleepUtils.java
@@ -4,7 +4,7 @@ import java.util.concurrent.ThreadLocalRandom;
 
 public final class SleepUtils {
 
-    public static void waitRandomTime(int sleepMs, int jitterMs) throws InterruptedException {
-        Thread.sleep(sleepMs + ThreadLocalRandom.current().nextInt(jitterMs));
+    public static void waitRandomTime(long sleepMs, long jitterMs) throws InterruptedException {
+        Thread.sleep(sleepMs + ThreadLocalRandom.current().nextLong(jitterMs));
     }
 }

--- a/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryErrorResponses.java
+++ b/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryErrorResponses.java
@@ -44,6 +44,7 @@ public class BigQueryErrorResponses {
   private static final String BAD_REQUEST_REASON = "badRequest";
   private static final String INVALID_REASON = "invalid";
   private static final String INVALID_QUERY_REASON = "invalidQuery";
+  private static final String JOB_INTERNAL_ERROR = "jobInternalError";
   private static final String NOT_FOUND_REASON = "notFound";
   private static final String QUOTA_EXCEEDED_REASON = "quotaExceeded";
   private static final String RATE_LIMIT_EXCEEDED_REASON = "rateLimitExceeded";
@@ -82,6 +83,11 @@ public class BigQueryErrorResponses {
     return BAD_REQUEST_CODE == exception.getCode()
         && exception.getError() == null
         && exception.getReason() == null;
+  }
+
+  public static boolean isJobInternalError(BigQueryException exception) {
+    return BAD_REQUEST_CODE == exception.getCode()
+            && JOB_INTERNAL_ERROR.equals(exception.getReason());
   }
 
   public static boolean isQuotaExceededError(BigQueryException exception) {

--- a/kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/MergeQueriesTest.java
+++ b/kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/MergeQueriesTest.java
@@ -20,10 +20,14 @@
 package com.wepay.kafka.connect.bigquery;
 
 import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryError;
+import com.google.cloud.bigquery.BigQueryException;
 import com.google.cloud.bigquery.Field;
 import com.google.cloud.bigquery.LegacySQLTypeName;
 import com.google.cloud.bigquery.Schema;
 import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TableResult;
+import com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException;
 import com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor;
 import com.wepay.kafka.connect.bigquery.write.batch.MergeBatches;
 import org.apache.kafka.connect.sink.SinkRecord;
@@ -38,8 +42,16 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
+import java.util.concurrent.CountDownLatch;
 
 import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyObject;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 @RunWith(MockitoJUnitRunner.class)
@@ -48,6 +60,8 @@ public class MergeQueriesTest {
   private static final String KEY = "kafkaKey";
 
   private static final int BATCH_NUMBER = 42;
+  private static final int BIGQUERY_RETRY = 3;
+  private static final int BIGQUERY_RETRY_WAIT = 1000;
   private static final TableId DESTINATION_TABLE = TableId.of("ds1", "t");
   private static final TableId INTERMEDIATE_TABLE = TableId.of("ds1", "t_tmp_6_uuid_epoch");
   private static final Schema INTERMEDIATE_TABLE_SCHEMA = constructIntermediateTable();
@@ -66,7 +80,7 @@ public class MergeQueriesTest {
 
   private MergeQueries mergeQueries(boolean insertPartitionTime, boolean upsert, boolean delete) {
     return new MergeQueries(
-        KEY, insertPartitionTime, upsert, delete, mergeBatches, executor, bigQuery, schemaManager, context
+      KEY, insertPartitionTime, upsert, delete, BIGQUERY_RETRY, BIGQUERY_RETRY_WAIT, mergeBatches, executor, bigQuery, schemaManager, context
     );
   }
 
@@ -334,6 +348,95 @@ public class MergeQueriesTest {
     mergeQueries(false, true, true).mergeFlush(INTERMEDIATE_TABLE);
 
     assertEquals(1, mergeBatches.incrementBatch(INTERMEDIATE_TABLE));
+  }
+
+  @Test
+  public void testBigQueryJobInternalErrorRetry() throws InterruptedException {
+    // Arrange
+    mergeBatches.addToBatch(TEST_SINK_RECORD, INTERMEDIATE_TABLE, new HashMap<>());
+
+    TableResult tableResultReponse = mock(TableResult.class);
+    BigQueryError jobInternalError = new BigQueryError("jobInternalError", null, "The job encountered an internal error during execution and was unable to complete successfully.");
+    when(bigQuery.query(anyObject()))
+            .thenThrow(new BigQueryException(400, "mock job internal error", jobInternalError))
+            .thenReturn(tableResultReponse);
+    when(mergeBatches.destinationTableFor(INTERMEDIATE_TABLE)).thenReturn(DESTINATION_TABLE);
+    when(mergeBatches.incrementBatch(INTERMEDIATE_TABLE)).thenReturn(0);
+    when(mergeBatches.prepareToFlush(INTERMEDIATE_TABLE, 0)).thenReturn(true);
+    CountDownLatch latch = new CountDownLatch(1);
+    doAnswer(invocation -> {
+      Runnable runnable = invocation.getArgument(0);
+      runnable.run();
+      latch.countDown();
+      return null;
+    }).when(executor).execute(any());
+    MergeQueries mergeQueries = spy(mergeQueries(false, true, true));
+
+    // Act
+    mergeQueries.mergeFlush(INTERMEDIATE_TABLE);
+
+    // Assert
+    latch.await();
+    verify(bigQuery, times(3)).query(anyObject());
+  }
+
+  @Test
+  public void testBigQueryInvalidQueryErrorRetry() throws InterruptedException {
+    // Arrange
+    mergeBatches.addToBatch(TEST_SINK_RECORD, INTERMEDIATE_TABLE, new HashMap<>());
+
+
+    TableResult tableResultReponse = mock(TableResult.class);
+    BigQueryError jobInternalError = new BigQueryError("invalidQuery", null, "Could not serialize access to table my_table due to concurrent update");
+    when(bigQuery.query(anyObject()))
+            .thenThrow(new BigQueryException(400, "mock invalid query", jobInternalError))
+            .thenReturn(tableResultReponse);
+    when(mergeBatches.destinationTableFor(INTERMEDIATE_TABLE)).thenReturn(DESTINATION_TABLE);
+    when(mergeBatches.incrementBatch(INTERMEDIATE_TABLE)).thenReturn(0);
+    when(mergeBatches.prepareToFlush(INTERMEDIATE_TABLE, 0)).thenReturn(true);
+    CountDownLatch latch = new CountDownLatch(1);
+    doAnswer(invocation -> {
+      Runnable runnable = invocation.getArgument(0);
+      runnable.run();
+      latch.countDown();
+      return null;
+    }).when(executor).execute(any());
+    MergeQueries mergeQueries = mergeQueries(false, true, true);
+
+    // Act
+    mergeQueries.mergeFlush(INTERMEDIATE_TABLE);
+
+    // Assert
+    latch.await();
+    verify(bigQuery, times(3)).query(anyObject());
+  }
+
+
+  @Test(expected = BigQueryConnectException.class)
+  public void testBigQueryRetryExceeded() throws InterruptedException {
+    // Arrange
+    mergeBatches.addToBatch(TEST_SINK_RECORD, INTERMEDIATE_TABLE, new HashMap<>());
+
+    BigQueryError jobInternalError = new BigQueryError("invalidQuery", null, "Could not serialize access to table my_table due to concurrent update");
+    when(bigQuery.query(anyObject()))
+      .thenThrow(new BigQueryException(400, "mock invalid query", jobInternalError));
+    when(mergeBatches.destinationTableFor(INTERMEDIATE_TABLE)).thenReturn(DESTINATION_TABLE);
+    when(mergeBatches.incrementBatch(INTERMEDIATE_TABLE)).thenReturn(0);
+    when(mergeBatches.prepareToFlush(INTERMEDIATE_TABLE, 0)).thenReturn(true);
+    CountDownLatch latch = new CountDownLatch(1);
+    doAnswer(invocation -> {
+      Runnable runnable = invocation.getArgument(0);
+      runnable.run();
+      latch.countDown();
+      return null;
+    }).when(executor).execute(any());
+    MergeQueries mergeQueries = mergeQueries(false, true, true);
+
+    // Act
+    mergeQueries.mergeFlush(INTERMEDIATE_TABLE);
+
+    //Assert
+    latch.await();
   }
 
   private String table(TableId table) {

